File: src/MixNetFx/Facesso.GenericControls/ucLabourValueListView.vb
Imports Facesso.Data

Public Class ucLabourValueListView
    Inherits System.Windows.Forms.ListView

    Private myAutoGroup As Boolean
    Private myLabourValueInfoCollection As LabourValueInfoCollection
    Private myLabourValueSortOrder As LabourValuesSortOrder
    Private myMaxDigitsLabourValueNo As Byte
    Private myMaxDigitsCostCenterNo As Byte
    Private myDataTable As DataTable
    Private myLastSortColumn As String

    Sub New()
        MyBase.New()
        Me.FullRowSelect = True
        Me.HideSelection = False
        Me.View = Windows.Forms.View.Details
        myLabourValueSortOrder = LabourValuesSortOrder.LabourValueNumber
        Me.Columns.Add("Arbeitswert-Nr.:", -2)
        Me.Columns.Add("Arbeitswert-Name/Tätigkeits-Beschreibung:", -2)
        Me.Columns.Add("Kostenstellenname:", -2)
        Me.Columns.Add("/nummer:", -2)
        Me.Columns.Add("Basiswertname", -2)
        Me.Columns.Add("Basiswert", -2)
        Me.Columns.Add("Einheit", -2)
        myLastSortColumn = "LabourValueNumber"
        myAutoGroup = True
    End Sub

    Public Property LabourValueSortOrder() As LabourValuesSortOrder
        Get
            Return myLabourValueSortOrder
        End Get
        Set(ByVal value As LabourValuesSortOrder)
            myLabourValueSortOrder = value
            rebuildList()
        End Set
    End Property

    Public Property LabourValues() As LabourValueInfoCollection
        Get
            Return myLabourValueInfoCollection
        End Get
        Set(ByVal value As LabourValueInfoCollection)
            myLabourValueInfoCollection = value
            If value IsNot Nothing Then
                SetMaxDigits()
            End If
            AssignToDataTable()
            rebuildList()
        End Set
    End Property

    Public ReadOnly Property FirstSelectedLabourValue() As LabourValueInfo
        Get
            If Me.SelectedIndices.Count = 0 Then
                Return Nothing
            End If
            Return Me.LabourValues(New ActiveDev.IntKey(Integer.Parse(Me.SelectedItems(0).Name)))
        End Get
    End Property

    Public ReadOnly Property SelectedLabourValues() As LabourValueInfoCollection
        Get
            Dim locLvic As New LabourValueInfoCollection
            For Each locLvi As ListViewItem In Me.SelectedItems
                locLvic.Add(Me.LabourValues(New ActiveDev.IntKey(Integer.Parse(locLvi.Name))))
            Next
            Return locLvic
        End Get
    End Property

    Public Sub SelectLabourValue(ByVal labourValue As LabourValueInfo, ByVal EnsureVisible As Boolean)
        For Each locLvi As ListViewItem In Me.SelectedItems
            If Integer.Parse(locLvi.Name) = labourValue.IDLabourValue Then
                locLvi.Selected = True
                If EnsureVisible Then
                    locLvi.EnsureVisible()
                End If
            End If
        Next
    End Sub

    Public Property AutoGroup() As Boolean
        Get
            Return myAutoGroup
        End Get
        Set(ByVal value As Boolean)
            myAutoGroup = value
            rebuildList()
        End Set
    End Property

    Private Sub rebuildList()
        Me.BeginUpdate()
        Me.Items.Clear()
        Me.Groups.Clear()
        If myDataTable Is Nothing Then
            Me.EndUpdate()
            Return
        End If

        myDataTable.DefaultView.Sort = LabourValueSortOrder.ToString & ", " & myLastSortColumn

        Dim locLastItem As ListViewItem = Nothing
        Dim locCurrentGroup As ListViewGroup = Nothing
        Dim LastDelta As Integer = -1

        For Each locRow As DataRowView In myDataTable.DefaultView
            Dim locItem As New ListViewItem(locRow("LabourValueNumber").ToString)
            locItem.SubItems.Add(locRow("LabourValueName").ToString)
            locItem.SubItems.Add(locRow("CostCenterName").ToString)
            locItem.SubItems.Add(locRow("CostCenterNo").ToString)
            locItem.SubItems.Add(locRow("BaseValueSynonym").ToString)
            locItem.SubItems.Add(locRow("TeHMin").ToString)
            locItem.SubItems.Add(locRow("Dimension").ToString)
            locItem.Name = locRow("IDLabourValue").ToString
            Me.Items.Add(locItem)

            'Die Gruppen bilden, falls es sich um die dafür richtige Sortierung handelt
            If AutoGroup Then
                If Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueName Or _
                    Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterName Or _
                    Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueNumber Or _
                    Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterNo Then
                    If Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueNumber Then
                        locItem.Tag = CInt(locRow("LabourValueNumber"))
                    ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueName Then
                        locItem.Tag = AscW(locRow("LabourValueName").ToString)
                    ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterNo Then
                        locItem.Tag = CInt(locRow("CostCenterNo"))
                    ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterName Then
                        locItem.Tag = CStr(locRow("CostCenterName"))
                    End If

                    If locCurrentGroup Is Nothing Then
                        If Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueNumber Then
                            locCurrentGroup = New ListViewGroup("Arbeitswerte ab Nummer:" & locRow("LabourValueNumber").ToString)
                            Me.Groups.Add(locCurrentGroup)
                        ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueName Then
                            Dim locCharValue As Integer = AscW(locRow("LabourValueName").ToString)
                            locCurrentGroup = New ListViewGroup("Arbeitswerte alphabetisch:" & ChrW(locCharValue))
                            Me.Groups.Add(locCurrentGroup)
                        ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterName Then
                            locCurrentGroup = New ListViewGroup("Arbeitswerte mit Kostenstellen namens:" & locRow("CostCenterName").ToString)
                            Me.Groups.Add(locCurrentGroup)
                        ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterNo Then
                            locCurrentGroup = New ListViewGroup("Arbeitswerte ab Kostenstellennr:" & locRow("CostCenterNo").ToString)
                            Me.Groups.Add(locCurrentGroup)
                        End If
                    End If

                    If locLastItem IsNot Nothing Then
                        If Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueNumber Then
                            If LastDelta > -1 Then
                                If LastDelta <> (CInt(locItem.Tag) - CInt(locLastItem.Tag)) Then
                                    locCurrentGroup = New ListViewGroup("Arbeitswerte ab Nummer:" & locRow("LabourValueNumber").ToString)
                                    Me.Groups.Add(locCurrentGroup)
                                    GoTo Label
                                End If
                            End If
                        ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueName Then
                            Dim locCharValue As Integer = AscW(locRow("LabourValueName").ToString)
                            If locCharValue <> CInt(locLastItem.Tag) Then
                                locCurrentGroup = New ListViewGroup("Arbeitswerte alphabetisch:" & ChrW(locCharValue))
                                Me.Groups.Add(locCurrentGroup)
                            End If
                        ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterName Then
                            If locRow("CostCenterName").ToString <> CStr(locLastItem.Tag) Then
                                locCurrentGroup = New ListViewGroup("Arbeitswerte mit Kostenstellen namens:" & locRow("CostCenterName").ToString)
                                Me.Groups.Add(locCurrentGroup)
                            End If
                        ElseIf Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterNo Then
                            If CInt(locItem.Tag) <> CInt(locLastItem.Tag) Then
                                locCurrentGroup = New ListViewGroup("Arbeitswerte ab Kostenstellennr:" & locRow("CostCenterNo").ToString)
                                Me.Groups.Add(locCurrentGroup)
                                GoTo Label
                            End If
                        End If

                        If Me.LabourValueSortOrder = LabourValuesSortOrder.LabourValueNumber Or _
                            Me.LabourValueSortOrder = LabourValuesSortOrder.CostCenterNo Then
                            LastDelta = CInt(locItem.Tag) - CInt(locLastItem.Tag)
                        End If
                    End If
Label:
                    locItem.Group = locCurrentGroup
                    locLastItem = locItem
                End If
            End If
        Next

        Me.Columns(0).Width = -2
        Me.Columns(1).Width = -2
        Me.Columns(2).Width = -2
        Me.Columns(3).Width = -2
        Me.Columns(4).Width = -1
        Me.Columns(5).Width = -1
        Me.EndUpdate()
    End Sub

    Private Sub AssignToDataTable()
        If myLabourValueInfoCollection Is Nothing Then
            myDataTable = Nothing
            Return
        End If
        myDataTable = New DataTable()
        With myDataTable.Columns
            .Add("IDLabourValue", GetType(Integer))
            .Add("LabourValueNumber", GetType(String))
            .Add("LabourValueName", GetType(String))
            .Add("CostCenterName", GetType(String))
            .Add("CostCenterNo", GetType(String))
            .Add("BaseValueSynonym", GetType(String))
            .Add("TeHMin", GetType(String))
            .Add("Dimension", GetType(String))
        End With

        For Each locLbi As LabourValueInfo In myLabourValueInfoCollection
            Dim locTc As DataRow = myDataTable.NewRow()
            locTc.Item("IDLabourValue") = locLbi.IDLabourValue
            locTc.Item("LabourValueNumber") = locLbi.LabourValueNumber.ToString(New String("0"c, myMaxDigitsLabourValueNo))
            locTc.Item("LabourValueName") = locLbi.LabourValueName
            locTc.Item("CostCenterName") = locLbi.CostCenterName
            locTc.Item("CostCenterNo") = locLbi.CostCenterNo.ToString(New String("0"c, myMaxDigitsCostCenterNo))
            locTc.Item("BaseValueSynonym") = locLbi.BaseValueSynonym
            locTc.Item("TeHMin") = locLbi.TeHMin.ToString("#,##0" & IIf(locLbi.BaseValuePrecision = 0, "", "." & New String("0"c, locLbi.BaseValuePrecision)).ToString)
            locTc.Item("Dimension") = locLbi.TeHMin
            myDataTable.Rows.Add(locTc)
        Next
    End Sub

    'TODO: Ausformulieren - Ermittelt die höchste Anzahl der Ziffern in der Liste
    Private Sub SetMaxDigits()
        myMaxDigitsCostCenterNo = 0
        myMaxDigitsLabourValueNo = 0
        For Each locLbi As LabourValueInfo In myLabourValueInfoCollection
            If locLbi.CostCenterNo.ToString.Length > myMaxDigitsCostCenterNo Then
                myMaxDigitsCostCenterNo = CByte(locLbi.CostCenterNo.ToString.Length)
            End If
            If locLbi.LabourValueNumber.ToString.Length > myMaxDigitsLabourValueNo Then
                myMaxDigitsLabourValueNo = CByte(locLbi.LabourValueNumber.ToString.Length)
            End If
        Next
    End Sub

    Protected Overrides Sub OnColumnClick(ByVal e As System.Windows.Forms.ColumnClickEventArgs)
        MyBase.OnColumnClick(e)
        myLastSortColumn = Me.LabourValueSortOrder.ToString
        Me.LabourValueSortOrder = DirectCast([Enum].ToObject(GetType(LabourValuesSortOrder), e.Column), LabourValuesSortOrder)
    End Sub
End Class

Public Enum LabourValuesSortOrder
    LabourValueNumber
    LabourValueName
    CostCenterName
    CostCenterNo
    BaseValueSynonym
    TeHMin
    Dimension
End Enum

File: src/NetFxCSharp/Facesso.GenericControls/ucLabourValueListView.cs
using Facesso.Data;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.GenericControls
{
    public class ucLabourValueListView : System.Windows.Forms.ListView
    {
        private bool myAutoGroup;
        private LabourValueInfoCollection myLabourValueInfoCollection;
        private LabourValuesSortOrder myLabourValueSortOrder;
        private byte myMaxDigitsLabourValueNo;
        private byte myMaxDigitsCostCenterNo;
        private DataTable myDataTable;
        private string myLastSortColumn;
        public ucLabourValueListView() : base()
        {
            this.FullRowSelect = true;
            this.HideSelection = false;
            this.View = System.Windows.Forms.View.Details;
            myLabourValueSortOrder = LabourValuesSortOrder.LabourValueNumber;
            this.Columns.Add("Arbeitswert-Nr.:", -2);
            this.Columns.Add("Arbeitswert-Name/Tätigkeits-Beschreibung:", -2);
            this.Columns.Add("Kostenstellenname:", -2);
            this.Columns.Add("/nummer:", -2);
            this.Columns.Add("Basiswertname", -2);
            this.Columns.Add("Basiswert", -2);
            this.Columns.Add("Einheit", -2);
            myLastSortColumn = "LabourValueNumber";
            myAutoGroup = true;
        }

        public LabourValuesSortOrder LabourValueSortOrder
        {
            get
            {
                return myLabourValueSortOrder;
            }

            set
            {
                myLabourValueSortOrder = value;
                rebuildList();
            }
        }

        public LabourValueInfoCollection LabourValues
        {
            get
            {
                return myLabourValueInfoCollection;
            }

            set
            {
                myLabourValueInfoCollection = value;
                if (value != null)
                {
                    SetMaxDigits();
                }

                AssignToDataTable();
                rebuildList();
            }
        }

        public LabourValueInfo FirstSelectedLabourValue
        {
            get
            {
                if (this.SelectedIndices.Count == 0)
                {
                    return null;
                }

                return this.LabourValues[new ActiveDev.IntKey(int.Parse(this.SelectedItems[0].Name))];
            }
        }

        public LabourValueInfoCollection SelectedLabourValues
        {
            get
            {
                LabourValueInfoCollection locLvic = new LabourValueInfoCollection();
                foreach (ListViewItem locLvi in this.SelectedItems)
                {
                    locLvic.Add(this.LabourValues[new ActiveDev.IntKey(int.Parse(locLvi.Name))]);
                }

                return locLvic;
            }
        }

        public void SelectLabourValue(LabourValueInfo labourValue, bool EnsureVisible)
        {
            foreach (ListViewItem locLvi in this.SelectedItems)
            {
                if (int.Parse(locLvi.Name) == labourValue.IDLabourValue)
                {
                    locLvi.Selected = true;
                    if (EnsureVisible)
                    {
                        locLvi.EnsureVisible();
                    }
                }
            }
        }

        public bool AutoGroup
        {
            get
            {
                return myAutoGroup;
            }

            set
            {
                myAutoGroup = value;
                rebuildList();
            }
        }

        private void rebuildList()
        {
            this.BeginUpdate();
            this.Items.Clear();
            this.Groups.Clear();
            if (myDataTable == null)
            {
                this.EndUpdate();
                return;
            }

            myDataTable.DefaultView.Sort = LabourValueSortOrder.ToString() + ", " + myLastSortColumn;
            ListViewItem locLastItem = null;
            ListViewGroup locCurrentGroup = null;
            int LastDelta = -1;
            foreach (DataRowView locRow in myDataTable.DefaultView)
            {
                ListViewItem locItem = new ListViewItem(locRow["LabourValueNumber"].ToString());
                locItem.SubItems.Add(locRow["LabourValueName"].ToString());
                locItem.SubItems.Add(locRow["CostCenterName"].ToString());
                locItem.SubItems.Add(locRow["CostCenterNo"].ToString());
                locItem.SubItems.Add(locRow["BaseValueSynonym"].ToString());
                locItem.SubItems.Add(locRow["TeHMin"].ToString());
                locItem.SubItems.Add(locRow["Dimension"].ToString());
                locItem.Name = locRow["IDLabourValue"].ToString();
                this.Items.Add(locItem);
                //Die Gruppen bilden, falls es sich um die dafür richtige Sortierung handelt
                if (AutoGroup)
                {
                    if (this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueName | this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterName | this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueNumber | this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterNo)
                    {
                        if (this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueNumber)
                        {
                            locItem.Tag = System.Convert.ToInt32(locRow["LabourValueNumber"]);
                        }
                        else if (this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueName)
                        {
                            locItem.Tag = Microsoft.VisualBasic.Strings.AscW(locRow["LabourValueName"].ToString());
                        }
                        else if (this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterNo)
                        {
                            locItem.Tag = System.Convert.ToInt32(locRow["CostCenterNo"]);
                        }
                        else if (this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterName)
                        {
                            locItem.Tag = System.Convert.ToString(locRow["CostCenterName"]);
                        }

                        if (locCurrentGroup == null)
                        {
                            if (this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueNumber)
                            {
                                locCurrentGroup = new ListViewGroup("Arbeitswerte ab Nummer:" + locRow["LabourValueNumber"].ToString());
                                this.Groups.Add(locCurrentGroup);
                            }
                            else if (this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueName)
                            {
                                int locCharValue = Microsoft.VisualBasic.Strings.AscW(locRow["LabourValueName"].ToString());
                                locCurrentGroup = new ListViewGroup("Arbeitswerte alphabetisch:" + Microsoft.VisualBasic.Strings.ChrW(locCharValue));
                                this.Groups.Add(locCurrentGroup);
                            }
                            else if (this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterName)
                            {
                                locCurrentGroup = new ListViewGroup("Arbeitswerte mit Kostenstellen namens:" + locRow["CostCenterName"].ToString());
                                this.Groups.Add(locCurrentGroup);
                            }
                            else if (this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterNo)
                            {
                                locCurrentGroup = new ListViewGroup("Arbeitswerte ab Kostenstellennr:" + locRow["CostCenterNo"].ToString());
                                this.Groups.Add(locCurrentGroup);
                            }
                        }

                        if (locLastItem != null)
                        {
                            if (this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueNumber)
                            {
                                if (LastDelta > -1)
                                {
                                    if (LastDelta != (System.Convert.ToInt32(locItem.Tag) - System.Convert.ToInt32(locLastItem.Tag)))
                                    {
                                        locCurrentGroup = new ListViewGroup("Arbeitswerte ab Nummer:" + locRow["LabourValueNumber"].ToString());
                                        this.Groups.Add(locCurrentGroup);
                                        goto Label;
                                    }
                                }
                            }
                            else if (this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueName)
                            {
                                int locCharValue = Microsoft.VisualBasic.Strings.AscW(locRow["LabourValueName"].ToString());
                                if (locCharValue != System.Convert.ToInt32(locLastItem.Tag))
                                {
                                    locCurrentGroup = new ListViewGroup("Arbeitswerte alphabetisch:" + Microsoft.VisualBasic.Strings.ChrW(locCharValue));
                                    this.Groups.Add(locCurrentGroup);
                                }
                            }
                            else if (this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterName)
                            {
                                if (locRow["CostCenterName"].ToString() != System.Convert.ToString(locLastItem.Tag))
                                {
                                    locCurrentGroup = new ListViewGroup("Arbeitswerte mit Kostenstellen namens:" + locRow["CostCenterName"].ToString());
                                    this.Groups.Add(locCurrentGroup);
                                }
                            }
                            else if (this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterNo)
                            {
                                if (System.Convert.ToInt32(locItem.Tag) != System.Convert.ToInt32(locLastItem.Tag))
                                {
                                    locCurrentGroup = new ListViewGroup("Arbeitswerte ab Kostenstellennr:" + locRow["CostCenterNo"].ToString());
                                    this.Groups.Add(locCurrentGroup);
                                    goto Label;
                                }
                            }

                            if (this.LabourValueSortOrder == LabourValuesSortOrder.LabourValueNumber | this.LabourValueSortOrder == LabourValuesSortOrder.CostCenterNo)
                            {
                                LastDelta = System.Convert.ToInt32(locItem.Tag) - System.Convert.ToInt32(locLastItem.Tag);
                            }
                        }

                        Label:
                            locItem.Group = locCurrentGroup;
                        locLastItem = locItem;
                    }
                }
            }

            this.Columns[0].Width = -2;
            this.Columns[1].Width = -2;
            this.Columns[2].Width = -2;
            this.Columns[3].Width = -2;
            this.Columns[4].Width = -1;
            this.Columns[5].Width = -1;
            this.EndUpdate();
        }

        private void AssignToDataTable()
        {
            if (myLabourValueInfoCollection == null)
            {
                myDataTable = null;
                return;
            }

            myDataTable = new DataTable();
            {
                var __with0 = myDataTable.Columns;
                __with0.Add("IDLabourValue", typeof(int));
                __with0.Add("LabourValueNumber", typeof(string));
                __with0.Add("LabourValueName", typeof(string));
                __with0.Add("CostCenterName", typeof(string));
                __with0.Add("CostCenterNo", typeof(string));
                __with0.Add("BaseValueSynonym", typeof(string));
                __with0.Add("TeHMin", typeof(string));
                __with0.Add("Dimension", typeof(string));
            }

            foreach (LabourValueInfo locLbi in myLabourValueInfoCollection)
            {
                DataRow locTc = myDataTable.NewRow();
                locTc["IDLabourValue"] = locLbi.IDLabourValue;
                locTc["LabourValueNumber"] = locLbi.LabourValueNumber.ToString(new string ('0', myMaxDigitsLabourValueNo));
                locTc["LabourValueName"] = locLbi.LabourValueName;
                locTc["CostCenterName"] = locLbi.CostCenterName;
                locTc["CostCenterNo"] = locLbi.CostCenterNo.ToString(new string ('0', myMaxDigitsCostCenterNo));
                locTc["BaseValueSynonym"] = locLbi.BaseValueSynonym;
                locTc["TeHMin"] = locLbi.TeHMin.ToString("#,##0" + Microsoft.VisualBasic.Interaction.IIf(locLbi.BaseValuePrecision == 0, "", "." + new string ('0', locLbi.BaseValuePrecision)).ToString());
                locTc["Dimension"] = locLbi.TeHMin;
                myDataTable.Rows.Add(locTc);
            }
        }

        //TODO: Ausformulieren - Ermittelt die höchste Anzahl der Ziffern in der Liste
        private void SetMaxDigits()
        {
            myMaxDigitsCostCenterNo = 0;
            myMaxDigitsLabourValueNo = 0;
            foreach (LabourValueInfo locLbi in myLabourValueInfoCollection)
            {
                if (locLbi.CostCenterNo.ToString().Length > myMaxDigitsCostCenterNo)
                {
                    myMaxDigitsCostCenterNo = System.Convert.ToByte(locLbi.CostCenterNo.ToString().Length);
                }

                if (locLbi.LabourValueNumber.ToString().Length > myMaxDigitsLabourValueNo)
                {
                    myMaxDigitsLabourValueNo = System.Convert.ToByte(locLbi.LabourValueNumber.ToString().Length);
                }
            }
        }

        protected override void OnColumnClick(System.Windows.Forms.ColumnClickEventArgs e)
        {
            base.OnColumnClick(e);
            myLastSortColumn = this.LabourValueSortOrder.ToString();
            this.LabourValueSortOrder = ((LabourValuesSortOrder)Enum.ToObject(typeof(LabourValuesSortOrder), e.Column));
        }
    }

    public enum LabourValuesSortOrder
    {
        LabourValueNumber,
        LabourValueName,
        CostCenterName,
        CostCenterNo,
        BaseValueSynonym,
        TeHMin,
        Dimension,
    }
}